pkg/cloudprovider/azure.go

```diff
@@ -1,37 +1,354 @@
+	for _, networkInterface := range networkInterfaces {
+		// Assign the IP
+		ipConfigurations := *networkInterface.IPConfigurations
+		name := fmt.Sprintf("%s_%s", node.Name, ip.String())
+		ipc := ip.String()
+		untrue := false
+		newIPConfiguration := network.InterfaceIPConfiguration{
+			Name: &name,
+			InterfaceIPConfigurationPropertiesFormat: &network.InterfaceIPConfigurationPropertiesFormat{
+				PrivateIPAddress:                &ipc,
+				PrivateIPAllocationMethod:       network.Static,
+				Subnet:                          (*networkInterface.IPConfigurations)[0].Subnet,
+				Primary:                         &untrue,
+				LoadBalancerBackendAddressPools: (*networkInterface.IPConfigurations)[0].LoadBalancerBackendAddressPools,
+			},
+		}
+		ipConfigurations = append(ipConfigurations, newIPConfiguration)
+		networkInterface.IPConfigurations = &ipConfigurations
+		// Send the request
+		result, err := a.createOrUpdate(networkInterface)
+		if err != nil {
+			return err
+		}
+		return a.waitForCompletion(result)
+	}
 	return nil
 }
 
 func (a *Azure) ReleasePrivateIP(ip net.IP, node *corev1.Node) error {
-	return nil
-}
-
-func (a *Azure) WaitForResponse(interface{}) error {
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return err
+	}
+	networkInterfaces, err := a.getNetworkInterfaces(instance)
+	if err != nil {
+		return err
+	}
+	// Perform the operation against the first interface listed, which will be
+	// the primary interface (if it's defined as such) or the first one returned
+	// following the order Azure specifies.
+	for _, networkInterface := range networkInterfaces {
+		// Release the IP
+		keepIPConfiguration := []network.InterfaceIPConfiguration{}
+		ipAssigned := false
+		for _, ipConfiguration := range *networkInterface.IPConfigurations {
+			if assignedIP := net.ParseIP(*ipConfiguration.PrivateIPAddress); assignedIP != nil && !assignedIP.Equal(ip) {
+				keepIPConfiguration = append(keepIPConfiguration, ipConfiguration)
+			} else if assignedIP != nil && assignedIP.Equal(ip) {
+				ipAssigned = true
+			}
+		}
+		// Short-circuit if the IP never existed to begin with
+		if !ipAssigned {
+			return NonExistingIPError
+		}
+		networkInterface.IPConfigurations = &keepIPConfiguration
+		// Send the request
+		result, err := a.createOrUpdate(networkInterface)
+		if err != nil {
+			return err
+		}
+		return a.waitForCompletion(result)
+	}
 	return nil
 }
 
 func (a *Azure) GetNodeEgressIPConfiguration(node *corev1.Node) ([]*NodeEgressIPConfiguration, error) {
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return nil, err
+	}
+	networkInterfaces, err := a.getNetworkInterfaces(instance)
+	if err != nil {
+		return nil, err
+	}
+	// Perform the operation against the first interface listed, which will be
+	// the primary interface (if it's defined as such) or the first one returned
+	// following the order Azure specifies.
+	for _, networkInterface := range networkInterfaces {
```

Comment:
The Azure code was a bit off: I just realized it had the network interface validation baked inside `getInstance`. I moved that to `getNetworkInterfaces` as to have it be more explicit that when we reach this part of the code: we are guaranteed to have at least one NIC. 

The reason I chose to write the code like this was to minimize the change in case we were ever thinking to deal with more than just the primary NIC.

Reply:
I changed it explicitly dealing with only the first item in the slice: